File: 1214-two-sum-bsts/1214-two-sum-bsts.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traverse(TreeNode* node,unordered_set<int>& set,int target) {
        //Base Case
        if(!node) return;
        // Fill the set
        set.insert(target-node->val);
        //left and right recursive call
        traverse(node->left,set,target);
        traverse(node->right,set,target);
        
    }
    bool traverse1(TreeNode* node,unordered_set<int>& set) {
        //Base Case
        if(!node) return false;
        //check if set contains Y -> (target-x)
        if(set.find(node->val) != set.end()) return true;
        //else run left and righ recursive calls
        bool leftFound = traverse1(node->left,set);
        bool rightFound = traverse1(node->right,set);
        
        return leftFound || rightFound;
    }
    bool twoSumBSTs(TreeNode* root1, TreeNode* root2, int target) {
        unordered_set<int> set;
        traverse(root1,set,target);
        return traverse1(root2,set);
    }
};

File: 1214-two-sum-bsts/1214-two-sum-bsts.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void traverse(TreeNode node,int target,HashSet<Integer> set) {
        //Base Case
        if(node == null) return;
        // Fill the set
        set.add(target - node.val);
        //left and right recursive call
        traverse(node.left,target,set);
        traverse(node.right,target,set);
    }
    
    public boolean traverse1(TreeNode node,HashSet<Integer> set) {
        //Base Case
        if(node == null) return false;
        //check if set contains Y -> (target-x)
        if(set.contains(node.val))return true;
        //else run left and righ recursive calls
        boolean leftFound = traverse1(node.left,set);
        boolean rightFound = traverse1(node.right,set);
        
        return leftFound || rightFound;
    }
    
    public boolean twoSumBSTs(TreeNode root1, TreeNode root2, int target) {
        HashSet<Integer> set = new HashSet<>();
        traverse(root1,target,set);
        return traverse1(root2,set);
    }
}